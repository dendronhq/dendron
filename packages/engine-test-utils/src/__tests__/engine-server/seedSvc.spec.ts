import {
  DEngineClient,
  DVault,
  ERROR_STATUS,
  SeedConfig,
} from "@dendronhq/common-all";
import { DConfig, tmpDir } from "@dendronhq/common-server";
import { SeedInitMode, SeedService, SeedUtils } from "@dendronhq/engine-server";
import os from "os";
import path from "path";
import { runEngineTestV5 } from "../../engine";
import { checkDir, checkFile, checkNotInDir, checkVaults } from "../../utils";
import { TestSeedUtils } from "../../utils/seed";

const BAD_SEED_ID = () => "dendron.no-exist";

async function createSeed({ engine }: { engine: DEngineClient }) {
  const tmp = tmpDir().name;
  const { registryFile, seedDict } = await TestSeedUtils.createSeedRegistry({
    engine,
    wsRoot: tmp,
  });
  const seedId = TestSeedUtils.defaultSeedId();
  return { registryFile, seedDict, seedId };
}

// Platform agnostic check file for a seed vault path
function getSeedVaultPathForCheckFile(seedId: string) {
  return path.join(`${seedId}`, `vault`).replace("\\", "\\\\");
}

// Skip on Windows for now until reliability issues can be fixed.
//TODO: Re-enable for Windows
const runTest = os.platform() === "win32" ? describe.skip : describe;

runTest("remove", () => {
  test("error: nothing to remove", async () => {
    await runEngineTestV5(
      async ({ wsRoot }) => {
        const id = BAD_SEED_ID();
        const seedService = new SeedService({ wsRoot });
        const resp = await seedService.removeSeed({ id });

        expect(resp.error?.status).toEqual(ERROR_STATUS.DOES_NOT_EXIST);
      },
      {
        expect,
      }
    );
  });

  test("ok: remove non-initialized seed", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        // create seed
        const {
          registryFile,
          seedDict,
          seedId: id,
        } = await createSeed({ engine });

        // add seed to config;
        const seed = seedDict[id];
        const seedService = new SeedService({ wsRoot, registryFile });
        await seedService.addSeedMetadata({ seed, wsRoot });

        // remove seed
        await seedService.removeSeed({ id });

        await checkFile({
          fpath: path.join(wsRoot, "dendron.yml"),
          snapshot: true,
        });
        await checkVaults(
          {
            wsRoot,
            vaults,
          },
          expect
        );
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });

  test("ok: remove initialized seed", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        // create seed
        const { registryFile, seedId: id } = await createSeed({ engine });

        // add seed to config;
        const seedService = new SeedService({ wsRoot, registryFile });
        await seedService.addSeed({ id });

        // remove seed
        await seedService.removeSeed({ id });

        const seedPath = SeedUtils.seed2Path({ wsRoot, id });
        await checkNotInDir(
          { fpath: path.dirname(seedPath), snapshot: true },
          id
        );
        expect(id).toMatchSnapshot();
        await checkFile({
          fpath: path.join(wsRoot, "dendron.yml"),
          snapshot: true,
        });
        await checkVaults(
          {
            wsRoot,
            vaults,
          },
          expect
        );
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });
});

runTest("add", () => {
  test("error: does not exist", async () => {
    await runEngineTestV5(
      async ({ wsRoot }) => {
        const id = "dendron.no-exist";

        const seedService = new SeedService({ wsRoot });
        const resp = await seedService.addSeed({ id });

        expect(resp.error?.status).toEqual(ERROR_STATUS.DOES_NOT_EXIST);
      },
      {
        expect,
      }
    );
  });

  test("ok: exists", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        const tmp = tmpDir().name;
        const { registryFile } = await TestSeedUtils.createSeedRegistry({
          engine,
          wsRoot: tmp,
        });
        const id = TestSeedUtils.defaultSeedId();

        const seedService = new SeedService({ wsRoot, registryFile });
        await seedService.addSeed({ id });

        // seed should be added
        await checkDir(
          { fpath: path.join(wsRoot, "seeds", id), snapshot: true },
          "dendron.yml",
          "seed.yml",
          "vault"
        );
        // seed.yml should not be present in the workspace
        await checkNotInDir(
          {
            fpath: wsRoot,
            snapshot: true,
          },
          "seed.yml"
        );
        await checkFile({
          fpath: path.join(wsRoot, "dendron.yml"),
          snapshot: true,
        });
        await checkFile(
          {
            fpath: path.join(wsRoot, "dendron.code-workspace"),
            snapshot: true,
          },
          getSeedVaultPathForCheckFile(id)
        );
        await checkVaults(
          {
            wsRoot,
            vaults: [{ fsPath: "vault", seed: id, name: id } as DVault].concat(
              vaults
            ),
          },
          expect
        );
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });

  test("ok: seed with site", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        const tmp = tmpDir().name;
        const { registryFile } = await TestSeedUtils.createSeedRegistry({
          engine,
          wsRoot: tmp,
          modifySeed: (seed) => {
            seed.site = {
              url: "https://foo.com",
              index: "foo",
            };
            return seed;
          },
        });
        const id = TestSeedUtils.defaultSeedId();

        const seedService = new SeedService({ wsRoot, registryFile });
        await seedService.addSeed({ id });

        // seed should be added
        await checkDir(
          { fpath: path.join(wsRoot, "seeds", id), snapshot: true },
          "dendron.yml",
          "seed.yml",
          "vault"
        );
        // seed.yml should not be present in the workspace
        await checkNotInDir(
          {
            fpath: wsRoot,
            snapshot: true,
          },
          "seed.yml"
        );
        await checkFile(
          {
            fpath: path.join(wsRoot, "dendron.yml"),
            snapshot: true,
          },
          "url: https://foo.com",
          "index: foo"
        );
        await checkFile(
          {
            fpath: path.join(wsRoot, "dendron.code-workspace"),
            snapshot: true,
          },
          getSeedVaultPathForCheckFile(id)
        );
        await checkVaults(
          {
            wsRoot,
            vaults: [{ fsPath: "vault", seed: id, name: id } as DVault].concat(
              vaults
            ),
          },
          expect
        );
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });
});

runTest("init", () => {
  const seed: SeedConfig = {
    id: "dendron.foo",
    name: "foo",
    publisher: "dendron",
    description: "some foo",
    license: "CC",
    root: "vault",
    repository: {
      type: "git",
      url: "",
    },
  };
  runTest("create workspace", () => {
    test(`basic`, async () => {
      await runEngineTestV5(
        async ({ wsRoot }) => {
          const seedService = new SeedService({ wsRoot });
          await seedService.init({
            seed,
            wsRoot,
            mode: SeedInitMode.CREATE_WORKSPACE,
          });

          await checkDir(
            {
              fpath: wsRoot,
              snapshot: true,
            },
            "dendron.yml",
            "seed.yml",
            "dendron.code-workspace",
            "vault"
          );
          await checkFile({
            fpath: path.join(wsRoot, "dendron.yml"),
            snapshot: true,
          });
          await checkFile({
            fpath: path.join(wsRoot, "seed.yml"),
            snapshot: true,
          });
          await checkVaults(
            {
              wsRoot,
              vaults: [
                {
                  fsPath: "vault",
                },
              ],
            },
            expect
          );
          expect(DConfig.getOrCreate(wsRoot).seeds).toEqual(undefined);
        },
        {
          expect,
          preSetupHook: async () => {},
          vaults: [],
        }
      );
    });
  });

  runTest("convert workspace", () => {
    const mode = SeedInitMode.CONVERT_WORKSPACE;

    const runInit = async (opts: { engine: DEngineClient; wsRoot: string }) => {
      const seedService = new SeedService({ wsRoot: opts.wsRoot });
      const resp = await seedService.init({ seed, wsRoot: opts.wsRoot, mode });

      return resp.error;
    };

    test(`error: no vaults`, async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const error = await runInit({ engine, wsRoot });
          expect(error?.message).toContain(
            "workspace must have exactly one vault"
          );
          await checkDir(
            {
              fpath: wsRoot,
              snapshot: true,
            },
            "dendron.yml"
          );
          await checkFile({
            fpath: path.join(wsRoot, "dendron.yml"),
            snapshot: true,
          });
          expect(DConfig.getOrCreate(wsRoot).seeds).toEqual(undefined);
        },
        {
          expect,
          vaults: [],
        }
      );
    });

    test(`error: wsVault`, async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const error = await runInit({ engine, wsRoot });
          expect(error?.message).toEqual("workspace vaults not supported");
        },
        {
          expect,
          vaults: [],
          workspaces: [
            {
              name: "fooWorkspace",
              vaults: [{ fsPath: "vault" }],
              remote: { type: "git", url: "" },
            },
          ],
        }
      );
    });

    test(`ok: diff root name`, async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const error = await runInit({ engine, wsRoot });
          expect(error).toBeUndefined();
          await checkDir(
            {
              fpath: wsRoot,
              snapshot: true,
            },
            "dendron.yml",
            "seed.yml"
          );
          await checkFile({
            fpath: path.join(wsRoot, "dendron.yml"),
            snapshot: true,
          });
          await checkFile(
            {
              fpath: path.join(wsRoot, "seed.yml"),
              snapshot: true,
            },
            "root: fooVault"
          );
        },
        {
          expect,
          vaults: [{ fsPath: "fooVault" }],
        }
      );
    });
  });
});

runTest("helpers", () => {
  test("ok: isSeedInWorkspace false", async () => {
    await runEngineTestV5(
      async ({ wsRoot }) => {
        const id = "dendron.no-exist";

        const seedService = new SeedService({ wsRoot });
        expect(seedService.isSeedInWorkspace(id)).toBeFalsy();
      },
      {
        expect,
      }
    );
  });

  test("ok: getSeedsInWorkspace empty", async () => {
    await runEngineTestV5(
      async ({ wsRoot }) => {
        const seedService = new SeedService({ wsRoot });
        const seedsInWS = seedService.getSeedsInWorkspace();

        expect(seedsInWS.length).toEqual(0);
      },
      {
        expect,
      }
    );
  });

  test("ok: isSeedInWorkspace true", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot }) => {
        const tmp = tmpDir().name;
        const { registryFile } = await TestSeedUtils.createSeedRegistry({
          engine,
          wsRoot: tmp,
        });
        const id = TestSeedUtils.defaultSeedId();

        const seedService = new SeedService({ wsRoot, registryFile });
        await seedService.addSeed({ id });

        expect(seedService.isSeedInWorkspace(id)).toBeTruthy();
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });

  test("ok: getSeedsInWorkspace non-empty", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot }) => {
        const tmp = tmpDir().name;
        const { registryFile } = await TestSeedUtils.createSeedRegistry({
          engine,
          wsRoot: tmp,
        });
        const id = TestSeedUtils.defaultSeedId();

        const seedService = new SeedService({ wsRoot, registryFile });
        await seedService.addSeed({ id });

        const seedsInWS = seedService.getSeedsInWorkspace();
        expect(seedsInWS.length).toEqual(1);
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });
});
