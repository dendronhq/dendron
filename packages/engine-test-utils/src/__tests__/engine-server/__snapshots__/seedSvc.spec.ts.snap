// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add ok: exists 1`] = `".git .gitignore dendron.code-workspace dendron.yml seed.yml vault"`;

exports[`add ok: exists 2`] = `"dendron.code-workspace dendron.yml hooks seeds vault1 vault2 vault3"`;

exports[`add ok: exists 3`] = `
"version: 5
dev:
    enablePreviewV2: true
commands:
    lookup:
        note:
            selectionMode: extract
            confirmVaultOnCreate: true
            vaultSelectionModeOnCreate: smart
            leaveTrace: false
            bubbleUpCreateNew: true
            fuzzThreshold: 0.2
    randomNote: {}
    insertNoteLink:
        aliasMode: none
        enableMultiSelect: false
    insertNoteIndex:
        enableMarker: false
    copyNoteLink:
        aliasMode: title
    templateHierarchy: template
workspace:
    vaults:
        -
            fsPath: vault
            seed: dendron.foo
            name: dendron.foo
        -
            fsPath: vault1
        -
            fsPath: vault2
        -
            fsPath: vault3
            name: vaultThree
    journal:
        dailyDomain: daily
        name: journal
        dateFormat: y.MM.dd
        addBehavior: childOfDomain
    scratch:
        name: scratch
        dateFormat: y.MM.dd.HHmmss
        addBehavior: asOwnDomain
    task:
        name: task
        dateFormat: y.MM.dd
        addBehavior: asOwnDomain
        statusSymbols:
            '': ' '
            wip: w
            done: x
            assigned: a
            moved: m
            blocked: b
            delegated: l
            dropped: d
            pending: 'y'
        taskCompleteStatus:
            - done
            - x
        prioritySymbols:
            H: high
            M: medium
            L: low
        todoIntegration: false
        createTaskSelectionType: selection2link
    graph:
        zoomSpeed: 1
        createStub: false
    enableAutoCreateOnDefinition: false
    enableXVaultWikiLink: false
    enableRemoteVaultInit: true
    enableUserTags: true
    enableHashTags: true
    workspaceVaultSyncMode: noCommit
    enableAutoFoldFrontmatter: false
    enableEditorDecorations: true
    maxPreviewsCached: 10
    maxNoteLength: 204800
    enableFullHierarchyNoteTitle: false
    seeds:
        dendron.foo: {}
preview:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enablePrettyRefs: true
    enableKatex: true
    automaticallyShowPreview: false
publishing:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enablePrettyRefs: true
    enableKatex: true
    copyAssets: true
    siteHierarchies:
        - root
    writeStubs: false
    siteRootDir: docs
    seo:
        title: Dendron
        description: Personal Knowledge Space
    github:
        enableEditLink: true
        editLinkText: Edit this page on GitHub
        editBranch: main
        editViewMode: tree
    enableSiteLastModified: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enableRandomlyColoredTags: true
    enableTaskNotes: true
    enablePrettyLinks: true
    searchMode: lookup
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault1
            - vault2
            - vaultThree
            - dendron.foo
"
`;

exports[`add ok: exists 4`] = `
"{
    \\"folders\\": [
        {
            \\"path\\": \\"seeds/dendron.foo/vault\\",
            \\"name\\": \\"dendron.foo\\"
        },
        {
            \\"path\\": \\"vault1\\"
        },
        {
            \\"path\\": \\"vault2\\"
        },
        {
            \\"path\\": \\"vault3\\",
            \\"name\\": \\"vaultThree\\"
        }
    ],
    \\"settings\\": {},
    \\"extensions\\": {}
}"
`;

exports[`add ok: seed with site 1`] = `".git .gitignore dendron.code-workspace dendron.yml seed.yml vault"`;

exports[`add ok: seed with site 2`] = `"dendron.code-workspace dendron.yml hooks seeds vault1 vault2 vault3"`;

exports[`add ok: seed with site 3`] = `
"version: 5
dev:
    enablePreviewV2: true
commands:
    lookup:
        note:
            selectionMode: extract
            confirmVaultOnCreate: true
            vaultSelectionModeOnCreate: smart
            leaveTrace: false
            bubbleUpCreateNew: true
            fuzzThreshold: 0.2
    randomNote: {}
    insertNoteLink:
        aliasMode: none
        enableMultiSelect: false
    insertNoteIndex:
        enableMarker: false
    copyNoteLink:
        aliasMode: title
    templateHierarchy: template
workspace:
    vaults:
        -
            fsPath: vault
            seed: dendron.foo
            name: dendron.foo
        -
            fsPath: vault1
        -
            fsPath: vault2
        -
            fsPath: vault3
            name: vaultThree
    journal:
        dailyDomain: daily
        name: journal
        dateFormat: y.MM.dd
        addBehavior: childOfDomain
    scratch:
        name: scratch
        dateFormat: y.MM.dd.HHmmss
        addBehavior: asOwnDomain
    task:
        name: task
        dateFormat: y.MM.dd
        addBehavior: asOwnDomain
        statusSymbols:
            '': ' '
            wip: w
            done: x
            assigned: a
            moved: m
            blocked: b
            delegated: l
            dropped: d
            pending: 'y'
        taskCompleteStatus:
            - done
            - x
        prioritySymbols:
            H: high
            M: medium
            L: low
        todoIntegration: false
        createTaskSelectionType: selection2link
    graph:
        zoomSpeed: 1
        createStub: false
    enableAutoCreateOnDefinition: false
    enableXVaultWikiLink: false
    enableRemoteVaultInit: true
    enableUserTags: true
    enableHashTags: true
    workspaceVaultSyncMode: noCommit
    enableAutoFoldFrontmatter: false
    enableEditorDecorations: true
    maxPreviewsCached: 10
    maxNoteLength: 204800
    enableFullHierarchyNoteTitle: false
    seeds:
        dendron.foo:
            site:
                url: https://foo.com
                index: foo
preview:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enablePrettyRefs: true
    enableKatex: true
    automaticallyShowPreview: false
publishing:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enablePrettyRefs: true
    enableKatex: true
    copyAssets: true
    siteHierarchies:
        - root
    writeStubs: false
    siteRootDir: docs
    seo:
        title: Dendron
        description: Personal Knowledge Space
    github:
        enableEditLink: true
        editLinkText: Edit this page on GitHub
        editBranch: main
        editViewMode: tree
    enableSiteLastModified: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enableRandomlyColoredTags: true
    enableTaskNotes: true
    enablePrettyLinks: true
    searchMode: lookup
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault1
            - vault2
            - vaultThree
            - dendron.foo
"
`;

exports[`add ok: seed with site 4`] = `
"{
    \\"folders\\": [
        {
            \\"path\\": \\"seeds/dendron.foo/vault\\",
            \\"name\\": \\"dendron.foo\\"
        },
        {
            \\"path\\": \\"vault1\\"
        },
        {
            \\"path\\": \\"vault2\\"
        },
        {
            \\"path\\": \\"vault3\\",
            \\"name\\": \\"vaultThree\\"
        }
    ],
    \\"settings\\": {},
    \\"extensions\\": {}
}"
`;

exports[`init convert workspace error: no vaults 1`] = `"dendron.code-workspace dendron.yml hooks"`;

exports[`init convert workspace error: no vaults 2`] = `
"version: 5
dev:
    enablePreviewV2: true
commands:
    lookup:
        note:
            selectionMode: extract
            confirmVaultOnCreate: true
            vaultSelectionModeOnCreate: smart
            leaveTrace: false
            bubbleUpCreateNew: true
            fuzzThreshold: 0.2
    randomNote: {}
    insertNoteLink:
        aliasMode: none
        enableMultiSelect: false
    insertNoteIndex:
        enableMarker: false
    copyNoteLink:
        aliasMode: title
    templateHierarchy: template
workspace:
    vaults: []
    journal:
        dailyDomain: daily
        name: journal
        dateFormat: y.MM.dd
        addBehavior: childOfDomain
    scratch:
        name: scratch
        dateFormat: y.MM.dd.HHmmss
        addBehavior: asOwnDomain
    task:
        name: task
        dateFormat: y.MM.dd
        addBehavior: asOwnDomain
        statusSymbols:
            '': ' '
            wip: w
            done: x
            assigned: a
            moved: m
            blocked: b
            delegated: l
            dropped: d
            pending: 'y'
        taskCompleteStatus:
            - done
            - x
        prioritySymbols:
            H: high
            M: medium
            L: low
        todoIntegration: false
        createTaskSelectionType: selection2link
    graph:
        zoomSpeed: 1
        createStub: false
    enableAutoCreateOnDefinition: false
    enableXVaultWikiLink: false
    enableRemoteVaultInit: true
    enableUserTags: true
    enableHashTags: true
    workspaceVaultSyncMode: noCommit
    enableAutoFoldFrontmatter: false
    enableEditorDecorations: true
    maxPreviewsCached: 10
    maxNoteLength: 204800
    enableFullHierarchyNoteTitle: false
preview:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enablePrettyRefs: true
    enableKatex: true
    automaticallyShowPreview: false
publishing:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enablePrettyRefs: true
    enableKatex: true
    copyAssets: true
    siteHierarchies:
        - root
    writeStubs: false
    siteRootDir: docs
    seo:
        title: Dendron
        description: Personal Knowledge Space
    github:
        enableEditLink: true
        editLinkText: Edit this page on GitHub
        editBranch: main
        editViewMode: tree
    enableSiteLastModified: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enableRandomlyColoredTags: true
    enableTaskNotes: true
    enablePrettyLinks: true
    searchMode: lookup
"
`;

exports[`init convert workspace ok: diff root name 1`] = `"dendron.code-workspace dendron.yml fooVault hooks seed.yml"`;

exports[`init convert workspace ok: diff root name 2`] = `
"version: 5
dev:
    enablePreviewV2: true
commands:
    lookup:
        note:
            selectionMode: extract
            confirmVaultOnCreate: true
            vaultSelectionModeOnCreate: smart
            leaveTrace: false
            bubbleUpCreateNew: true
            fuzzThreshold: 0.2
    randomNote: {}
    insertNoteLink:
        aliasMode: none
        enableMultiSelect: false
    insertNoteIndex:
        enableMarker: false
    copyNoteLink:
        aliasMode: title
    templateHierarchy: template
workspace:
    vaults:
        -
            fsPath: fooVault
    journal:
        dailyDomain: daily
        name: journal
        dateFormat: y.MM.dd
        addBehavior: childOfDomain
    scratch:
        name: scratch
        dateFormat: y.MM.dd.HHmmss
        addBehavior: asOwnDomain
    task:
        name: task
        dateFormat: y.MM.dd
        addBehavior: asOwnDomain
        statusSymbols:
            '': ' '
            wip: w
            done: x
            assigned: a
            moved: m
            blocked: b
            delegated: l
            dropped: d
            pending: 'y'
        taskCompleteStatus:
            - done
            - x
        prioritySymbols:
            H: high
            M: medium
            L: low
        todoIntegration: false
        createTaskSelectionType: selection2link
    graph:
        zoomSpeed: 1
        createStub: false
    enableAutoCreateOnDefinition: false
    enableXVaultWikiLink: false
    enableRemoteVaultInit: true
    enableUserTags: true
    enableHashTags: true
    workspaceVaultSyncMode: noCommit
    enableAutoFoldFrontmatter: false
    enableEditorDecorations: true
    maxPreviewsCached: 10
    maxNoteLength: 204800
    enableFullHierarchyNoteTitle: false
preview:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enablePrettyRefs: true
    enableKatex: true
    automaticallyShowPreview: false
publishing:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enablePrettyRefs: true
    enableKatex: true
    copyAssets: true
    siteHierarchies:
        - root
    writeStubs: false
    siteRootDir: docs
    seo:
        title: Dendron
        description: Personal Knowledge Space
    github:
        enableEditLink: true
        editLinkText: Edit this page on GitHub
        editBranch: main
        editViewMode: tree
    enableSiteLastModified: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enableRandomlyColoredTags: true
    enableTaskNotes: true
    enablePrettyLinks: true
    searchMode: lookup
    duplicateNoteBehavior:
        action: useVault
        payload:
            - fooVault
"
`;

exports[`init convert workspace ok: diff root name 3`] = `
"id: dendron.foo
name: foo
publisher: dendron
description: some foo
license: CC
root: fooVault
repository:
    type: git
    url: ''
"
`;

exports[`init create workspace basic 1`] = `".gitignore dendron.code-workspace dendron.yml hooks seed.yml vault"`;

exports[`init create workspace basic 2`] = `
"version: 5
dev:
    enablePreviewV2: true
commands:
    lookup:
        note:
            selectionMode: extract
            confirmVaultOnCreate: true
            vaultSelectionModeOnCreate: smart
            leaveTrace: false
            bubbleUpCreateNew: true
            fuzzThreshold: 0.2
    randomNote: {}
    insertNoteLink:
        aliasMode: none
        enableMultiSelect: false
    insertNoteIndex:
        enableMarker: false
    copyNoteLink:
        aliasMode: title
    templateHierarchy: template
workspace:
    vaults:
        -
            fsPath: vault
    journal:
        dailyDomain: daily
        name: journal
        dateFormat: y.MM.dd
        addBehavior: childOfDomain
    scratch:
        name: scratch
        dateFormat: y.MM.dd.HHmmss
        addBehavior: asOwnDomain
    task:
        name: task
        dateFormat: y.MM.dd
        addBehavior: asOwnDomain
        statusSymbols:
            '': ' '
            wip: w
            done: x
            assigned: a
            moved: m
            blocked: b
            delegated: l
            dropped: d
            pending: 'y'
        taskCompleteStatus:
            - done
            - x
        prioritySymbols:
            H: high
            M: medium
            L: low
        todoIntegration: false
        createTaskSelectionType: selection2link
    graph:
        zoomSpeed: 1
        createStub: false
    enableAutoCreateOnDefinition: false
    enableXVaultWikiLink: false
    enableRemoteVaultInit: true
    enableUserTags: true
    enableHashTags: true
    workspaceVaultSyncMode: noCommit
    enableAutoFoldFrontmatter: false
    enableEditorDecorations: true
    maxPreviewsCached: 10
    maxNoteLength: 204800
    enableFullHierarchyNoteTitle: false
preview:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enablePrettyRefs: true
    enableKatex: true
    automaticallyShowPreview: false
publishing:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enablePrettyRefs: true
    enableKatex: true
    copyAssets: true
    siteHierarchies:
        - root
    writeStubs: false
    siteRootDir: docs
    seo:
        title: Dendron
        description: Personal Knowledge Space
    github:
        enableEditLink: true
        editLinkText: Edit this page on GitHub
        editBranch: main
        editViewMode: tree
    enableSiteLastModified: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enableRandomlyColoredTags: true
    enableTaskNotes: true
    enablePrettyLinks: true
    searchMode: search
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault
"
`;

exports[`init create workspace basic 3`] = `
"id: dendron.foo
name: foo
publisher: dendron
description: some foo
license: CC
root: vault
repository:
    type: git
    url: ''
"
`;

exports[`remove ok: remove initialized seed 1`] = `""`;

exports[`remove ok: remove initialized seed 2`] = `"dendron.foo"`;

exports[`remove ok: remove initialized seed 3`] = `
"version: 5
dev:
    enablePreviewV2: true
commands:
    lookup:
        note:
            selectionMode: extract
            confirmVaultOnCreate: true
            vaultSelectionModeOnCreate: smart
            leaveTrace: false
            bubbleUpCreateNew: true
            fuzzThreshold: 0.2
    randomNote: {}
    insertNoteLink:
        aliasMode: none
        enableMultiSelect: false
    insertNoteIndex:
        enableMarker: false
    copyNoteLink:
        aliasMode: title
    templateHierarchy: template
workspace:
    vaults:
        -
            fsPath: vault1
        -
            fsPath: vault2
        -
            fsPath: vault3
            name: vaultThree
    journal:
        dailyDomain: daily
        name: journal
        dateFormat: y.MM.dd
        addBehavior: childOfDomain
    scratch:
        name: scratch
        dateFormat: y.MM.dd.HHmmss
        addBehavior: asOwnDomain
    task:
        name: task
        dateFormat: y.MM.dd
        addBehavior: asOwnDomain
        statusSymbols:
            '': ' '
            wip: w
            done: x
            assigned: a
            moved: m
            blocked: b
            delegated: l
            dropped: d
            pending: 'y'
        taskCompleteStatus:
            - done
            - x
        prioritySymbols:
            H: high
            M: medium
            L: low
        todoIntegration: false
        createTaskSelectionType: selection2link
    graph:
        zoomSpeed: 1
        createStub: false
    enableAutoCreateOnDefinition: false
    enableXVaultWikiLink: false
    enableRemoteVaultInit: true
    enableUserTags: true
    enableHashTags: true
    workspaceVaultSyncMode: noCommit
    enableAutoFoldFrontmatter: false
    enableEditorDecorations: true
    maxPreviewsCached: 10
    maxNoteLength: 204800
    enableFullHierarchyNoteTitle: false
    seeds: {}
preview:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enablePrettyRefs: true
    enableKatex: true
    automaticallyShowPreview: false
publishing:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enablePrettyRefs: true
    enableKatex: true
    copyAssets: true
    siteHierarchies:
        - root
    writeStubs: false
    siteRootDir: docs
    seo:
        title: Dendron
        description: Personal Knowledge Space
    github:
        enableEditLink: true
        editLinkText: Edit this page on GitHub
        editBranch: main
        editViewMode: tree
    enableSiteLastModified: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enableRandomlyColoredTags: true
    enableTaskNotes: true
    enablePrettyLinks: true
    searchMode: lookup
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault1
            - vault2
            - vaultThree
"
`;

exports[`remove ok: remove non-initialized seed 1`] = `
"version: 5
dev:
    enablePreviewV2: true
commands:
    lookup:
        note:
            selectionMode: extract
            confirmVaultOnCreate: true
            vaultSelectionModeOnCreate: smart
            leaveTrace: false
            bubbleUpCreateNew: true
            fuzzThreshold: 0.2
    randomNote: {}
    insertNoteLink:
        aliasMode: none
        enableMultiSelect: false
    insertNoteIndex:
        enableMarker: false
    copyNoteLink:
        aliasMode: title
    templateHierarchy: template
workspace:
    vaults:
        -
            fsPath: vault1
        -
            fsPath: vault2
        -
            fsPath: vault3
            name: vaultThree
    journal:
        dailyDomain: daily
        name: journal
        dateFormat: y.MM.dd
        addBehavior: childOfDomain
    scratch:
        name: scratch
        dateFormat: y.MM.dd.HHmmss
        addBehavior: asOwnDomain
    task:
        name: task
        dateFormat: y.MM.dd
        addBehavior: asOwnDomain
        statusSymbols:
            '': ' '
            wip: w
            done: x
            assigned: a
            moved: m
            blocked: b
            delegated: l
            dropped: d
            pending: 'y'
        taskCompleteStatus:
            - done
            - x
        prioritySymbols:
            H: high
            M: medium
            L: low
        todoIntegration: false
        createTaskSelectionType: selection2link
    graph:
        zoomSpeed: 1
        createStub: false
    enableAutoCreateOnDefinition: false
    enableXVaultWikiLink: false
    enableRemoteVaultInit: true
    enableUserTags: true
    enableHashTags: true
    workspaceVaultSyncMode: noCommit
    enableAutoFoldFrontmatter: false
    enableEditorDecorations: true
    maxPreviewsCached: 10
    maxNoteLength: 204800
    enableFullHierarchyNoteTitle: false
    seeds: {}
preview:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enablePrettyRefs: true
    enableKatex: true
    automaticallyShowPreview: false
publishing:
    enableFMTitle: true
    enableNoteTitleForLink: true
    enablePrettyRefs: true
    enableKatex: true
    copyAssets: true
    siteHierarchies:
        - root
    writeStubs: false
    siteRootDir: docs
    seo:
        title: Dendron
        description: Personal Knowledge Space
    github:
        enableEditLink: true
        editLinkText: Edit this page on GitHub
        editBranch: main
        editViewMode: tree
    enableSiteLastModified: true
    enableFrontmatterTags: true
    enableHashesForFMTags: false
    enableRandomlyColoredTags: true
    enableTaskNotes: true
    enablePrettyLinks: true
    searchMode: lookup
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault1
            - vault2
            - vaultThree
"
`;
